Tests/MongoSwiftSyncTests/MongoCollectionTests.swift

```diff
@@ -589,6 +589,105 @@ final class MongoCollectionTests: MongoSwiftTestCase {
             .to(throwError(errorType: MongoError.InvalidArgumentError.self))
     }
 
+    func testHintInPreviousVersion() throws {
+        try withTestNamespace { client, _, collection in
+            let requirements = TestRequirement(
+                maxServerVersion: ServerVersion(major: 4, minor: 2)
+            )
+            let unmetRequirement = try client.getUnmetRequirement(requirements)
+            guard unmetRequirement == nil else {
+                printSkipMessage(testName: self.name, unmetRequirement: unmetRequirement!)
+                return
+            }
+
+            // Drivers must raise an error for servers < 4.2.
+            let invalidOpts1 = FindOneAndReplaceOptions(hint: ["nonexistant_index": 1])
+            let invalidOpts2 = FindOneAndReplaceOptions(hint: "nonexistant_index")
+            let invalidOpts3 = FindOneAndUpdateOptions(hint: ["nonexistant_index": 1])
+            let invalidOpts4 = FindOneAndUpdateOptions(hint: "nonexistant_index")
+            expect(try collection.findOneAndReplace(filter: [:], replacement: [:], options: invalidOpts1))
+                .to(throwError(errorType: MongoError.InternalError.self))
```

Comment:
Generally speaking, we shouldn't be `InternalError`s, since those are reserved for buggy behavior / system failure. We'll likely need to add some more support for parsing the error that libmongoc is returning here to return a more specific one.

From looking at the libmongoc source for this, it seems to yield a `bson_error_t` with domain `MONGOC_ERROR_COMMAND` and code `MONGOC_ERROR_PROTOCOL_BAD_WIRE_VERSION` on < 4.2 when a hint is spcified. Looking at `parseMongocError` in `MongoError.swift`, it looks like we don't handle this particular combination, hence why this is currently throwing `InternalError`. We should update that method to return a `CompatibilityError`, like we do for the`(MONGOC_ERROR_PROTOCOL, MONGOC_ERROR_PROTOCOL_BAD_WIRE_VERSION)` combination, and then these failures should be more accurate.

Reply:
done